Allow combining lifetime scope tags

chaoaretasty opened the pull request:
Autofac's lifetimeScope registration methods such as `InstancePerRequest` and `InstancePerMatchingLifetimeScope` take a `params object[] lifetimeScopeTags` at the end allowing easy combining of allowed scopes. I've added this parameter to `InstancePerBackgroundJob`.

This lets people call `InstancePerBackgroundJob(Autofac.Core.Lifetime.MatchingScopeLifetimeTags.RequestLifetimeScopeTag)` allowing for proper registration of a component for either webrequests or background jobs, as well as allowing people to bring in their own tags if needed.

odinserj commented:
Maybe it is better to use `AutofacJobActivator.LifetimeScopeTag` (public since 2.0.1) for this completely custom scenario, leaving `InstancePerRequest` and `InstancePerBackgroundJob` methods untouched? We can document this field in the readme file.

chaoaretasty commented:
Regarding the previous PR I would agree it's probably baking in too much for this scenario, however this one is just aligning the API with Autofac's equivalents. Both `InstancePerRequest` and `InstancePerMatchingLifetimeScope` take `params object[] lifetimeScopeTags` so I'd say it's still a worthwhile change, and as a `params` it would not be a breaking change to implementers.

Either way I would agree the readme needs updating as well.

odinserj commented:
So I agree with this pull request, if it makes `InstancePerBackgroundJob` to be equivalent to regular Autofac methods. But there is one note. When you modify a method signature, you are making a breaking change for current users. There is an alternative without making a breaking change – add an overload with `params` instead.

kelostrada commented:
hey, some tests could be appreciated as well?

chaoaretasty commented:
Can do it as an overload if you want though as it's a `params` it's only going to be breaking to people using reflection on it and even then only when looking at the signature with it (i honestly can't see a situation where anyone would want to do this). Your call though on that.

Good point, I'll get round to adding tests this evening.

odinserj commented:
Some users are doing drop-in replacement. When a project is compiled against an old method, there will be a run-time error. It is always better to not to introduce breaking changes, where possible.

chaoaretasty commented:
No problems, not something I had considered. I'll update this evening along with tests

chaoaretasty commented:
Changed to an overload and added two new tests. No asserts as we are testing that it doesn't throw when resolved from either lifetime scope (the case if we had chained the registrations).

odinserj commented:
@kelostrada, would you like to review the changes?

kelostrada commented on HangFire.Autofac/RegistrationExtensions.cs:
```diff
@@ -25,8 +26,31 @@ public static class RegistrationExtensions
             InstancePerBackgroundJob<TLimit, TActivatorData, TStyle>(
             [NotNull] this IRegistrationBuilder<TLimit, TActivatorData, TStyle> registration)
         {
-            if (registration == null) throw new ArgumentNullException("registration");
```
There should still be a check for argument here:
`if (registration == null) throw new ArgumentNullException("registration");`

kelostrada commented on HangFire.Autofac.Tests/AutofacJobActivatorTests.cs:
```diff
@@ -178,6 +178,34 @@ public void InstanceRegisteredWith_InstancePerJob_IsDisposedOnScopeDisposal()
+
+            using (var scope = container.BeginLifetimeScope(alternateLifetimeScopeTag))
+            {
+                var instance = scope.Resolve(typeof(object));
```
shouldn't you check if resolved object instance is the same or different object depending on scope type than another object instance you resolved?

chaoaretasty replied:
There's already tests that Hangfire.Autofac properly registers with its scope tag when needed and that the activator correctly instantiates new lifetime scopes. The only job of this change is to combine any incoming scope tags with hangfire's when registering, which is tested by being able to resolve from scopes of either lifetime and doesn't throw for a missing tag.

I could add that test in too if you want but you're really writing tests for Autofac itself at that point that its lifetime scopes work as it claims..

kelostrada replied:
okay you are right :) good point, I have no other questions I think

chaoaretasty commented:
Readme updated. I removed the previous HTTP Warning section as there was overlap and replaced it with a warning about mixing lifetime scopes. If some of the previous content needs to be put back I'm not quite sure how to best word it to fit in.

kelostrada commented:
I like it, I think you can merge :)

odinserj commented:
Thanks guys and sorry for the delay! Merging and releasing.
